askama_derive/src/generator.rs

```diff
@@ -727,7 +727,41 @@ impl<'a> Generator<'a> {
                 args.len()
             )));
         }
-        for (expr, arg) in std::iter::zip(args, &def.args) {
+        let mut named_arguments = HashMap::new();
+        // Since named arguments can only be passed last, we only need to check if the last argument
+        // is a named one.
+        if let Some(Expr::NamedArgument(_, _)) = args.last() {
+            // First we check that all named arguments actually exist in the called item.
+            for arg in args.iter().rev() {
+                if let Expr::NamedArgument(arg_name, _) = arg {
+                    if !def.args.iter().any(|arg| arg == arg_name) {
+                        return Err(CompileError::from(format!(
+                            "no argument named `{arg_name}` in macro {name:?}"
+                        )));
+                    }
+                    named_arguments.insert(arg_name, arg);
+                } else {
+                    break;
+                }
+            }
+        }
+
+        // Handling both named and unnamed arguments requires to be careful of the named arguments
+        // order. To do so, we iterate through the macro defined arguments and then check if we have
+        // a named argument with this name:
+        //
+        // * If there is one, we add it and move to the next argument.
+        // * If there isn't one, then we pick the next argument (we can do it without checking
+        //   anything since named arguments are always last).
+        let mut index = 0;
```

Comment:
Instead of this index, I could use an iterator. As you prefer.

Reply:
This seems fine.